### DISCUSSION TITLE
How to mock the provider for Storybook and Jest?

### DISCUSSION THREAD

**Discussion by ldiego08** [wevm/wagmi#335] [Q&A]
Hi, I need to mock the `Provider` instance to use it on Storybook and Jest. Ideally, I shouldn't need to use an actual wallet to connect or send transactions. I tried using `wagmi-testing` but Storybook fails with missing node built-ins, which could mean `setupServer` is getting called even if I only import the `MockConnector`. Any recommendations on how to achieve this?

**Comment by tmm** [comment]
Check out how jest is set up in the source code. Nothing magical happening.

What build system are you using for Storybook? If you are using Vite (or other ES module set ups), you'll need to make sure the browser has access to node built-ins (like [this](https://github.com/tmm/wagmi/blob/main/examples/vite-react/index.html#L9-L15=)). You could also probably drop `CoinbaseWalletConnector` and `WalletConnectConnector` since they are the only parts of wagmi that rely on these built-ins (not my choice, something these two projects decided to use).

**Reply by ldiego08** [reply]
@tmm as recommended, I looked at how tests are set up. I see they're using the `MockConnector` class coming from `@wagmi/core`. It does require some extra configuration, namely copying the [`setupClient`](https://github.com/wagmi-dev/wagmi/blob/ba80a54f1c0df4166dbc272c012c671c6efbe7f4/packages/react/test/utils.ts#L23) and [`getProvider`](https://github.com/wagmi-dev/wagmi/blob/ba80a54f1c0df4166dbc272c012c671c6efbe7f4/packages/core/test/utils.ts#L20) (and all related code). I kind of got it to work, but hooks like `useBalance` are failing with a network error as the setup code uses `ethers.providers.StaticJsonRpcProvider` which will ultimately do a network call to `http://127.0.0.1:8545/`. My question would be, is there's a server being mocked at that address for wagmi tests? Or am I missing more config to mock network calls?

<img width="1507" alt="Screen Shot 2022-08-10 at 11 57 31" src="https://user-images.githubusercontent.com/271803/183983508-cf7ece64-312c-42fe-ab51-bbd9a439f75e.png">

**Reply by ldiego08** [reply]
I also noticed it uses `react-query`, so it makes sense it's making a network call?

**Reply by tmm** [reply]
Yup! We are running [Anvil](https://github.com/foundry-rs/foundry).
